### PR TITLE
v3.3.0

### PR DESCRIPTION
**Pull request by z64**
[Rendered Changelog](https://github.com/z64/discordrb/blob/release/3.3.0/CHANGELOG.md)

Feel free to review with feedback, corrections, things missing, etc.

Todo:
- [x] Bold "breaking" changes
- [x] General release notes, if we feel those are necessary (suggestions welcome)

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -5,7 +5,158 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
...
+- Support for category channels in `Server` and `Channel` ([#415](https://github.com/meew0/discordrb/pull/415), [#477](https://github.com/meew0/discordrb/pull/477), thanks @omnilord)
+- `CommandBot` and commands channel white-list can now be modified after creation ([#446](https://github.com/meew0/discordrb/pull/446), thanks @omnilord)
+- A `Role`'s `position` can now be sorted relative to other roles ([#445](https://github.com/meew0/discordrb/pull/445), thanks @mattantonelli)
+- The `return` keyword inside of commands can be used to return content to Discord ([#462](https://github.com/meew0/discordrb/pull/462), thanks @TrenchFroast)
+- `Emoji` now supplies `animated` ([#464](https://github.com/meew0/discordrb/pull/464), thanks @PixeLInc)
+- Additional instructions for installation of Ruby's devkit for Ruby 2.3+ ([#468](https://github.com/meew0/discordrb/pull/468), thanks @oct2pus)
+- `Server` API for retrieving a server's audit logs ([#353](https://github.com/meew0/discordrb/pull/353), thanks @Snazzah)
+- `EventContainer` methods for server role create, delete, and update events ([#494](https://github.com/meew0/discordrb/pull/494), thanks @Daniel-Worral, thanks @Daniel-Worrall)
```
Comment (connorshea) [inline CHANGELOG.md]:
```suggestion
- `EventContainer` methods for server role create, delete, and update events ([#494](https://github.com/meew0/discordrb/pull/494), thanks @Daniel-Worrall)
```

Double thanks

File: CHANGELOG.md [hunk 2]
```diff
@@ -5,7 +5,158 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
...
+- Additional objects and attributes for parsing embeds on messages ([#344](https://github.com/meew0/discordrb/pull/344), thanks @mattantonelli)
+- Methods for finding a members highest role, the role that is hoisting the member, or giving the member color ([#335](https://github.com/meew0/discordrb/pull/335), thanks @Snazzah)
+- API support for managing webhooks ([#356](https://github.com/meew0/discordrb/pull/356), thanks @Daniel-Worrall)
+- Support for reading and managing a channel's `nsfw` property ([#380](https://github.com/meew0/discordrb/pull/380))
+- The `:administrator` permissions value is aliased as `:administrate` ([#322](https://github.com/meew0/discordrb/pull/322))
+- Class methods on `Permissions` for easily building permissions bits values ([#322](https://github.com/meew0/discordrb/pull/322))
+- `Gateway#send_packet` and `Gateway#send_raw` methods to send custom data payloads to the gateway
+- Methods for reading `icon_url` and `proxy_icon_url` in `EmbedAuthor`
+- Methods for obtaining a server and channels invites ([#394](https://github.com/meew0/discordrb/pull/394))
+- Example of using awaits ([#370](https://github.com/meew0/discordrb/pull/370))
```
Comment (connorshea) [inline CHANGELOG.md]:
```suggestion
- Example of using `awaits` ([#370](https://github.com/meew0/discordrb/pull/370))
```

Not formatting it as code? git gud, scrub

File: CHANGELOG.md [hunk 3]
```diff
@@ -5,7 +5,158 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
...
+
+### Added
+
+- API methods to add and remove single member roles ([#310](https://github.com/meew0/discordrb/pull/310))
+- API methods and abstractions for listing available voice regions ([#311](https://github.com/meew0/discordrb/pull/311))
+- `Server` methods to prune members and to get the number of members available for pruning ([#315](https://github.com/meew0/discordrb/pull/315))
+- Methods for filtering the kinds of overwrites present on a channel ([#321](https://github.com/meew0/discordrb/pull/321))
+- `Channel#default_channel?`, for checking if a channel is the default channel of a server ([#320](https://github.com/meew0/discordrb/pull/320), thanks @Reaver01)
+- Method for returning a `Server`'s `@everyone` role
+- Reaction can now be serialized with `#to_s` to be used in `Message#react` more easily ([#342](https://github.com/meew0/discordrb/pull/342))
```
Comment (connorshea) [inline CHANGELOG.md]:
```suggestion
- Reactions can now be serialized with `#to_s` to be used in `Message#react` more easily ([#342](https://github.com/meew0/discordrb/pull/342))
```
smh amateur

File: CHANGELOG.md [hunk 4]
```diff
@@ -5,7 +5,158 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
...
+- Filled in permissions bit for viewing a server's audit log ([#349](https://github.com/meew0/discordrb/pull/349), thanks @Daniel-Worrall)
+- https://cdn.discordapp.com is now used as the base URL for CDN resources like avatars and server icons ([#358](https://github.com/meew0/discordrb/pull/358))
+- Reaction events raised from the bot's actions will respect `parse_self` ([#350](https://github.com/meew0/discordrb/pull/350), thanks @Daniel-Worrall)
+- `Webhooks::Embed#initialize` parses its `color` argument ([#364](https://github.com/meew0/discordrb/pull/364), thanks @Daniel-Worrall)
+- Webhook related events can now be matched on webhook ID ([#363](https://github.com/meew0/discordrb/pull/363), thanks @Daniel-Worrall)
+- Discord's default user avatar URLs will now be returned when applicable ([#375](https://github.com/meew0/discordrb/pull/375))
+- `Cache#find_user` can now find individual users if name and discriminator is given ([#384](https://github.com/meew0/discordrb/pull/384))
+- `ReactionEvent` provides server, and member if possible ([#351](https://github.com/meew0/discordrb/pull/351), thanks @Daniel-Worrall)
```
Comment (connorshea) [inline CHANGELOG.md]: Is this supposed to be "server and member if possible", as in like "server if possible and member if possible"? The comma isn't really grammatically correct regardless, but meh I'm not sure how to properly phrase that.

File: CHANGELOG.md (hunk 2 above)
Comment (z64) [reply CHANGELOG.md]: Well, no - awaits is the concept. The only "`awaits`" is `Bot#awaits`, which really isn't the scope of this comment

File: CHANGELOG.md [hunk 5]
```diff
@@ -5,7 +5,158 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
...
+- Installation instructions now include guides for installing with bundler ([#386](https://github.com/meew0/discordrb/pull/386), [#405](https://github.com/meew0/discordrb/pull/405), thanks @VxJasonxV and @PixeLInc)
```
Comment (connorshea) [inline CHANGELOG.md]:
```suggestion
- Installation instructions now include guides for installing with Bundler ([#386](https://github.com/meew0/discordrb/pull/386), [#405](https://github.com/meew0/discordrb/pull/405), thanks @VxJasonxV and @PixeLInc)
```

File: CHANGELOG.md [hunk 6]
```diff
@@ -5,7 +5,158 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
...
+- `default_channel` implementation is updated to reflect Discord changes ([#382](https://github.com/meew0/discordrb/pull/382), [#534](https://github.com/meew0/discordrb/pull/534))
+- Documentation around the conditions where our API returns `nil` is clarified ([#395](https://github.com/meew0/discordrb/pull/395), thanks @LikeLakers2)
+- Whenever possible, we update cached data about a `Server` returned to us from making changes to it
+- `Cache#server` now returns `nil` if a server is not found instead of raising an exception ([#424](https://github.com/meew0/discordrb/pull/424), thanks @soukouki)
+- `Bucket#rate_limited?` now accepts an `increment` value for weighted rate limits ([#427](https://github.com/meew0/discordrb/pull/427), thanks @Lavode)
+- `Server#bans` now returns `Array<ServerBan>`, which contains `reason` data in addition to the user banned ([#404](https://github.com/meew0/discordrb/pull/404))
+- `Channel#pune` now accepts a block that can be used to filter the messages to be pruned ([#421](https://github.com/meew0/discordrb/pull/421), thanks @snapcase)
```
Comment (connorshea) [inline CHANGELOG.md]: `pune`?

```suggestion
- `Channel#prune` now accepts a block that can be used to filter the messages to be pruned ([#421](https://github.com/meew0/discordrb/pull/421), thanks @snapcase)
```

File: CHANGELOG.md [hunk 7]
```diff
@@ -5,7 +5,158 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
...
+- `nonce` is accepted in `API::Channel.create_message` ([#414](https://github.com/meew0/discordrb/pull/414), thanks @PixeLInc)
+- Setters for new status options (`Bot#listening=`, `Bot#watching=`) ([#432](https://github.com/meew0/discordrb/pull/432), thanks @PixeLInc)
+- Documentation examples for sending a file ([#409](https://github.com/meew0/discordrb/pull/409))
+- Respondable implements `#send_embed` ([#420](https://github.com/meew0/discordrb/pull/420))
+- `Invite` now supplies `max_age` and `created_at`
+- `Invite` now supplies `member_count` and `online_member_count` ([#454](https://github.com/meew0/discordrb/pull/454), thanks @Snazzah)
+- `Server` methods for managing a server's embed (widget) settings ([#435](https://github.com/meew0/discordrb/pull/435))
+- Support for category channels in `Server` and `Channel` ([#415](https://github.com/meew0/discordrb/pull/415), [#477](https://github.com/meew0/discordrb/pull/477), thanks @omnilord)
+- `CommandBot` and commands channel white-list can now be modified after creation ([#446](https://github.com/meew0/discordrb/pull/446), thanks @omnilord)
```
Comment (connorshea) [inline CHANGELOG.md]: whitelist is one word

```suggestion
- `CommandBot` and commands channel whitelist can now be modified after creation ([#446](https://github.com/meew0/discordrb/pull/446), thanks @omnilord)
```

Comment (z64) [reply CHANGELOG.md]: Yeah, my spellcheck plugin flagged it.. I thought so.

File: CHANGELOG.md [hunk 8]
```diff
@@ -5,7 +5,158 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
 ## Unreleased
-[View diff for this release.](https://github.com/meew0/discordrb/compare/v3.2.1...HEAD)
+[View diff for this release.](https://github.com/meew0/discordrb/compare/v3.3.0...HEAD)
+
+## [3.3.0] - TBD
```
Comment (connorshea) [inline CHANGELOG.md]: Don't forget to update the date here when it's ready to merge.

File: CHANGELOG.md [hunk 9]
```diff
@@ -5,7 +5,158 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
...
+- Support for reading and managing a channel's `nsfw` property ([#380](https://github.com/meew0/discordrb/pull/380))
+- The `:administrator` permissions value is aliased as `:administrate` ([#322](https://github.com/meew0/discordrb/pull/322))
+- Class methods on `Permissions` for easily building permissions bits values ([#322](https://github.com/meew0/discordrb/pull/322))
+- `Gateway#send_packet` and `Gateway#send_raw` methods to send custom data payloads to the gateway
+- Methods for reading `icon_url` and `proxy_icon_url` in `EmbedAuthor`
+- Methods for obtaining a server and channels invites ([#394](https://github.com/meew0/discordrb/pull/394))
+- Example of using awaits ([#370](https://github.com/meew0/discordrb/pull/370))
+- Methods on `Member` for kicking and banning ([#404](https://github.com/meew0/discordrb/pull/404))
+- API method and abstraction for adding members to guilds with OAuth2 tokens ([#413](https://github.com/meew0/discordrb/pull/413))
+- Example of using a prefix proc ([#411](https://github.com/meew0/discordrb/pull/413))
```
Comment (connorshea) [inline CHANGELOG.md]: This PR doesn't seem to be the right one?

File: CHANGELOG.md [hunk 10]
```diff
@@ -5,7 +5,158 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
...
+### Changed
+
+- `Channel#make_invite` now accepts an argument to always return a unique invite code ([#312](https://github.com/meew0/discordrb/pull/312))
+- More of the API accepts objects that respond to `#resolve_id` ([#313](https://github.com/meew0/discordrb/pull/313), [#328](https://github.com/meew0/discordrb/pull/328), thanks @Likelakers2)
+- `Channel#history` and `API::Channel.messages` now accepts `around_id` ([#314](https://github.com/meew0/discordrb/pull/314))
+- `API::Server.prune_count` accepts `days` ([#315](https://github.com/meew0/discordrb/pull/315))
+- Methods for creating channels accept additional arguments ([#321](https://github.com/meew0/discordrb/pull/321))
+- `Channel` overwrite-related API now returns an `Overwrite` object ([#321](https://github.com/meew0/discordrb/pull/321))
+- Creating roles now accepts more parameters ([#323](https://github.com/meew0/discordrb/pull/323), thanks @Reaver01)
+- Ratelimits are now logged to a `:ratelimit` logging level and can be configured
```
Comment (connorshea) [inline CHANGELOG.md]:
```suggestion
- Rate Limits are now logged to a `:ratelimit` logging level and can be configured
```

**Comment by connorshea** [comment]
What would "General release notes" entail? 🤔

**Comment by z64** [comment]
See [3.0.0](https://github.com/meew0/discordrb/blob/master/CHANGELOG.md#300---2016-09-30) for example. Basically, any kind of guide to the most important points of the release; a "TL;DR"

File: CHANGELOG.md [hunk 11]
```diff
@@ -5,7 +5,158 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
...
+- Methods for managing a server's system channel ([#437](https://github.com/meew0/discordrb/pull/437), thanks @ldelelis)
+- Additional error code constants ([#419](https://github.com/meew0/discordrb/pull/419), thanks @LikeLakers2)
+- Commands can be created with a `:rescue` argument, to provide a message or callback when an unhandled exception is raised when executing the command ([#360](https://github.com/meew0/discordrb/pull/360))
+- Additional `Server` properties for verification levels, default message notification levels, and explicit content filter settings ([#414](https://github.com/meew0/discordrb/pull/414), thanks @PixeLInc)
+- `nonce` is accepted in `API::Channel.create_message` ([#414](https://github.com/meew0/discordrb/pull/414), thanks @PixeLInc)
+- Setters for new status options (`Bot#listening=`, `Bot#watching=`) ([#432](https://github.com/meew0/discordrb/pull/432), thanks @PixeLInc)
+- Documentation examples for sending a file ([#409](https://github.com/meew0/discordrb/pull/409))
+- Respondable implements `#send_embed` ([#420](https://github.com/meew0/discordrb/pull/420))
+- `Invite` now supplies `max_age` and `created_at`
```
Comment (connorshea) [inline CHANGELOG.md]: I can't find any commit for this?

File: CHANGELOG.md [hunk 12]
```diff
@@ -5,7 +5,158 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
...
+- The gemspec now includes a link to the changelog ([#515](https://github.com/meew0/discordrb/pull/515), thanks @ChallahuAkbar)
+- Commands can now be restricted by either `allowed_roles` and `required_roles` ([#469](https://github.com/meew0/discordrb/pull/469), thanks @elfenars)
+- `Bot#parse_mention` parses `Channel` mentions ([#525](https://github.com/meew0/discordrb/pull/525), thanks @estherbolik)
+- Support for Discord's `zlib-stream` gateway compression, as well as options to configure the compression mode in `Bot#initialize` ([#527](https://github.com/meew0/discordrb/pull/527), thanks @oct2pus for additional testing)
+- "Priority Speaker" permission bit ([#530](https://github.com/meew0/discordrb/pull/530), thanks @Chewsterchew)
+- Implemented `aliases` attribute in commands, for an improved alternative to "command copying" by passing an array to the command name ([#524](https://github.com/meew0/discordrb/pull/524))
+- Methods for managing a `Channel`'s slowmode settings ([#573](https://github.com/meew0/discordrb/pull/573), thanks @badBlackShark)
```
[the remark this answers is not in the document]
Comment (z64) [reply CHANGELOG.md]: It's `slowmode` in the desktop client and our API

File: CHANGELOG.md [hunk 13]
```diff
@@ -5,7 +5,158 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
...
+### Changed
+
+- `Channel#make_invite` now accepts an argument to always return a unique invite code ([#312](https://github.com/meew0/discordrb/pull/312))
+- More of the API accepts objects that respond to `#resolve_id` ([#313](https://github.com/meew0/discordrb/pull/313), [#328](https://github.com/meew0/discordrb/pull/328), thanks @Likelakers2)
+- `Channel#history` and `API::Channel.messages` now accepts `around_id` ([#314](https://github.com/meew0/discordrb/pull/314))
+- `API::Server.prune_count` accepts `days` ([#315](https://github.com/meew0/discordrb/pull/315))
+- Methods for creating channels accept additional arguments ([#321](https://github.com/meew0/discordrb/pull/321))
+- `Channel` overwrite-related API now returns an `Overwrite` object ([#321](https://github.com/meew0/discordrb/pull/321))
+- Creating roles now accepts more parameters ([#323](https://github.com/meew0/discordrb/pull/323), thanks @Reaver01)
+- Rate limits are now logged to a `:ratelimit` logging level and can be configured
```
[the remark this answers is not in the document]
Comment (z64) [reply CHANGELOG.md]: Yes

**Pull request merged**